src/tests/integration_tests/test_usage_tracker.py

```diff
@@ -0,0 +1,160 @@
+import os
+import threading
+
+import pytest
+
+from sherpa_ai.database.user_usage_tracker import UserUsageTracker
+from sherpa_ai.verbose_loggers.base import BaseVerboseLogger
+import shutil
+import sqlite3
+
+user_id = "test_id"
+
+def create_testing_db_file(db_path):
+    # Connect to the SQLite database (if it doesn't exist, it will be created)
+    conn = sqlite3.connect(db_path)
+    conn.close()
+    print(f"Database file '{db_path}' created successfully.")
+
+# Specify the path where you want to create the database file
+db_path = "./tests/data/test_usage_counter.db"
+
+# Call the function to create the database file
+create_testing_db_file(db_path)
+
+
+def test_usage_tracker_whitelist_add_and_get():
+    check = os.path.exists("test_usage.db")
+    if check:
+        os.remove("test_usage.db")
+    db = UserUsageTracker(db_name="test_usage.db", db_url="sqlite:///test_usage.db")
+    db.add_to_whitelist(user_id=user_id)
+    white_listed_id = db.get_all_whitelisted_ids()
+    assert user_id in white_listed_id
+
+
+def test_usage_checker():
+    check = os.path.exists("test_usage.db")
+    if check:
+        os.remove("test_usage.db")
+    db = UserUsageTracker(db_name="test_usage.db", db_url="sqlite:///test_usage.db" , )
+    db.max_daily_token = 2000
+    db.limit_time_size_in_hours = 0.1
+    check_usage = db.check_usage(token_amount="3000", user_id="jack")
+    assert check_usage["can_execute"] == False
+    check_usage = db.check_usage(token_amount="1000", user_id="jack")
+    assert check_usage["can_execute"] == True
+    check_usage = db.check_usage(token_amount="1000", user_id="jack")
+    assert check_usage["can_execute"] == True and check_usage["token-left"] == 1000
+    check_usage = db.check_usage(token_amount="1000", user_id="jack")
+    assert check_usage["can_execute"] == False and check_usage["token-left"] == 0
+
+
+def test_usage_reset_after_a_given_time():
+    check = os.path.exists("test_usage.db")
+    if check:
+        os.remove("test_usage.db")
+    db = UserUsageTracker(db_name="test_usage.db", db_url="sqlite:///test_usage.db")
+    db.max_daily_token = 2000
+    db.limit_time_size_in_hours = 0.001  # 3.6 seconds
+    db.check_usage(token_amount="2000", user_id="jack")
+    check_usage = db.check_usage(token_amount="10", user_id="jack")
+    assert check_usage["can_execute"] == False
+
+    # check if the usage tracker has been updated after 4 seconds
+    timer = threading.Timer(4.0, (lambda: ""))
```

Comment:
Never put delays in tests. They are error-prone because the delay duration is non-deterministic. Delays also slow down the test run and we want tests to be fast.

Instead, mock the time methods so that Python thinks the time has moved forward 4 seconds.
This library might help: https://pypi.org/project/freezegun/.  You can add it as a test dependency.
If you can make this approach work quickly then please proceed, otherwise comment out this test.

A cleaner approach would be to make as much of the code purely functional as possible by passing in values from the DB at the top of the function call chain. Then the lower-level functions that evaluate various logical conditions can be tested without mocking. But this requires a bigger refactor than we have time for.